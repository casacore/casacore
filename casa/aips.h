//# aips.h: Global initialization for standard types, etc.
//# Copyright (C) 1993-1999,2000,2002
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef CASA_AIPS_H
#define CASA_AIPS_H

//# Define compiler specific flags
#include <casacore/casa/aipsenv.h>

//# Define the standard types used by Casacore.
#include <casacore/casa/aipstype.h>

//# Define the extra non-standard types used by Casacore
#include <casacore/casa/aipsxtype.h>

//# Define the special Casacore macros
#include <casacore/casa/aipsdef.h>

namespace casa { //# NAMESPACE CASA - BEGIN

} //# NAMESPACE CASA - END

#endif
