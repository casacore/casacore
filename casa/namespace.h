#ifndef NAMESPACE_H
#define NAMESPACE_H 
using namespace casa;
#endif
