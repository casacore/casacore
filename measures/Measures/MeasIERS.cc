//# MeasIERS.cc: Interface to IERS tables
//# Copyright (C) 1996-2003,2007,2008
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

//# Includes
#include <measures/Measures/MeasIERS.h>
#include <casa/Arrays/Vector.h>
#include <casa/Exceptions/Error.h>
#include <casa/Logging/LogIO.h>
#include <casa/BasicMath/Math.h>
#include <casa/OS/Time.h>
#include <casa/OS/Path.h>
#include <casa/Quanta/Quantum.h>
#include <casa/Quanta/MVTime.h>
#include <casa/System/Aipsrc.h>
#include <casa/System/AipsrcValue.h>
#include <casa/BasicSL/String.h>

namespace casa { //# NAMESPACE CASA - BEGIN

#ifndef CASADATA
#define CASADATA "/usr/local"
#endif

//# Constants
const Double MeasIERS::INTV = 5;

//# Static data
uInt MeasIERS::predicttime_reg = 0;
uInt MeasIERS::notable_reg = 0;
uInt MeasIERS::forcepredict_reg = 0;

//# Member functions
Bool MeasIERS::get(Double &returnValue,
		   MeasIERS::Files file, 
		   MeasIERS::Types type, 
		   Double date) {
  returnValue = 0.0;
  if (!MeasIERS::predicttime_reg) {
    predicttime_reg = 
      AipsrcValue<Double>::registerRC(String("measures.measiers.d_predicttime"),
				      Unit("d"), Unit("d"),
				      MeasIERS::INTV);
    notable_reg = 
      AipsrcValue<Bool>::registerRC(String("measures.measiers.b_notable"),
				    False);
    forcepredict_reg = 
      AipsrcValue<Bool>::registerRC(String("measures.measiers.b_forcepredict"),
				    False);
  }
  if (AipsrcValue<Bool>::get(MeasIERS::notable_reg)) return True;
  Bool res = True;
  Double f;
  if (!dateNow) dateNow = Time().modifiedJulianDay();
  if (file == PREDICTED ||
      AipsrcValue<Bool>::get(MeasIERS::forcepredict_reg) ||
      (dateNow - date) <= 
      AipsrcValue<Double>::get(MeasIERS::predicttime_reg)) {
    res = False;			// do predict
  }
  MeasIERS::Files which = MeasIERS::MEASURED;
  if (res) {
    res = initMeas(which);
    if (res) {
      res = fillMeas(which, date);
    }
  }
  if (!res) {				// Retry for predicted
    which = MeasIERS::PREDICTED;
    if (!initMeas(which)) return False;
    if (!fillMeas(which, date)) {
      if (!msgDone) {
	LogIO os(LogOrigin("MeasIERS",
			   String("fillMeas(MeasIERS::Files, Double)"),
			   WHERE));
        Time now;       // current time
        
        if(date > now.modifiedJulianDay()){
          // People using times from the future are almost certainly simulating
          // data, and, even if they would be upset by the IERS table not being
          // available, there is not much they can do about it.
          os << LogIO::NORMAL3
             << "High precision Earth axis data is not yet available for requested JD "
             << date
             << LogIO::POST;
        }
        else{
          os << LogIO::NORMAL
             << "Requested JD " << date
             << " is outside the range of the IERS (Earth axis data) table." 	 
             << "\nCalculations will proceed with less precision"
             << LogIO::POST;
        }
	msgDone = True;
      }
      return False;
    }
  }
  // Interpolation fraction
  f = date - ldat[which][0];
  returnValue = ldat[which+N_Files][type]*f - ldat[which][type]*(f-1.0);
  return True;
}

Bool MeasIERS::initMeas(MeasIERS::Files which) {
  static const String names[MeasIERS::N_Types] = {
    "MJD",
    "x",
    "y",
    "dUT1",
    "LOD",
    "dPsi",
    "dEps",
    "Dx",
    "Dy",
    "DdUT1",
    "DLOD",
    "DdPsi",
    "DdEps"};
  static const String tplc[N_Files] = {"measures.ierseop97.directory",
                                       "measures.ierspredict.directory"};

  if (measFlag[which]) {
    ScopedMutexLock locker(theirMutex);
    if (measFlag[which]) {
      TableRecord kws;
      Double dt;
      String vs;
      Bool ok = True;
      if (!MeasIERS::getTable(MeasIERS::t[which], kws, row[which],
                              rfp[Int(which)], vs, dt, 
                              N_Types, names, tp[which],
                              tplc[which],
                              "geodetic")) {
        LogIO os(LogOrigin("MeasIERS",
                           String("initMeas(MeasIERS::Files)"),
                           WHERE));
        os << LogIO::NORMAL1
           << "Cannot read IERS (Earth axis data) table " << tp[which]
           << "\nCalculations will proceed with lower precision"
           << LogIO::POST;
        ok = False;
      }
      if (ok) {
        MeasIERS::openNote(&MeasIERS::closeMeas);
        if (!kws.isDefined("MJD0") || kws.asDouble("MJD0") < 10000) {
          ok = False;
        }
      }
      if (ok) {
        mjd0[which] = Int(kws.asDouble("MJD0"));
        Int n = t[which].nrow();
        row[which].get(n-1);
        if (*(rfp[Int(which)][0]) != mjd0[which] + n) { 
          ok = False;
        } else {
          mjdl[which] = mjd0[which] + n;
        }
      }
      if (!ok) {
        LogIO os(LogOrigin("MeasIERS",
                           String("initMeas(MeasIERS::Files)"),
                           WHERE));
        os << String("Corrupted IERS table ") + tp[which] << LogIO::EXCEPTION;
      }
      for (Int i = 0; i < MeasIERS::N_Types; i++) {
        for (Int j = 0; j < 2*MeasIERS::N_Files; j++) {
          ldat[j][i] = 0;
        }
      }
      measFlag[which] = False;
    }
  }
  return (! t[which].isNull());
}

void MeasIERS::closeMeas() {
  for (uInt i=0; i<N_Files; ++i) {
    if (!measFlag[i]) {
      ScopedMutexLock locker(theirMutex);
      if (!measFlag[i]) {
        if (! t[i].isNull()) {
          dateNow = 0.0;
          mjd0[i] = 0;
          mjdl[i] = 0;
          msgDone = False;
          t[i] = Table();
        }
        measFlag[i] = True;
      }
    }
  }
}

void MeasIERS::openNote(CLOSEFUN fun) {
  if (nNote <= sizeNote) {
    CLOSEFUN *tmp = 0;
    if (sizeNote > 0) {
      tmp = new CLOSEFUN[sizeNote];
      for (uInt i=0; i<sizeNote; ++i) tmp[i] = toclose[i];
      delete [] toclose; toclose = 0;
    }
    toclose = new CLOSEFUN[sizeNote+10];
    for (uInt i=0; i<sizeNote; ++i) toclose[i] = tmp[i];
    for (uInt i=sizeNote; i<sizeNote+10; ++i) toclose[i] = 0;
    sizeNote += 10;
    delete [] tmp; tmp = 0;
  }
  toclose[nNote++] = fun;
}

void MeasIERS::closeTables() {
  if (sizeNote > 0) {
    for (uInt i=sizeNote; i-1 != 0; --i) {
      if (toclose[i-1] != 0) toclose[i-1]();
      toclose[i-1] = 0;
    }
    delete [] toclose; toclose = 0;
    sizeNote = 0;
    nNote = 0;
  }
}

// Table handling
Bool MeasIERS::getTable(Table &table, TableRecord &kws, ROTableRow &row,
			RORecordFieldPtr<Double> rfp[],
			String &vs, Double &dt,
			Int N, const String rfn[],
			const String &name,
			const String &rc, const String &dir,
			const Table *tabin) {
  Table tab;
  Bool ok = findTab(tab, tabin, rc, dir, name);

  if(!ok)
    return false;            // findTab logs its own errors.
  
  LogIO os(LogOrigin("MeasIERS",
		     String("getTable(Table &, TableRecord &, "
			    "ROTableRow &, RORecordFieldPtr<Double> *, "
			    "String &vs, Double &dt, "
			    "Int N, const String *, const String &, "
			    "const String &, const String &)"),
		     WHERE));

  TableRecord ks(tab.keywordSet());

  ok = handle_keywords(dt, vs, ks, tab);
  
  ROTableRow rw(tab);
  if (ok) {
    // Check that the table is not missing any expected columns.
    for (Int i=0; i < N; i++) {
      if (!rw.record().isDefined(rfn[i])) {
	os << LogIO::SEVERE
	   << "Column " << rfn[i] << " is missing."
	   << LogIO::POST;
	ok = False;// break;
      }
    }
  }
  if (!ok) {
    os << name + " has an incompatible format."
       << "\nYou may want to notify the CASA system manager about it."
       << LogIO::EXCEPTION;
    return False;
  }
  table = tab;
  kws = ks;
  row = rw;
  for (Int i=0; i < N; i++)
    rfp[i] = RORecordFieldPtr<Double>(row.record(), rfn[i]);
  return True;
}

Bool MeasIERS::getTable(Table &table, TableRecord &kws, ROTableRow &row,
			Vector<RORecordFieldPtr<Double> >& rfp,
			String &vs, Double &dt,
			const Vector<String>& reqcols,
			Vector<String>& optcols,
			const String &name,
			const String &rc, const String &dir,
			const Table *tabin)
{
  Table tab;
  Bool ok = findTab(tab, tabin, rc, dir, name);

  if(!ok)
    return false;            // findTab logs its own errors.
  
  LogIO os(LogOrigin("MeasIERS", "getTable(Vector<String>& optcols)",
		     WHERE));

  TableRecord ks(tab.keywordSet());

  ok = handle_keywords(dt, vs, ks, tab);
  
  ROTableRow rw(tab);
  if(ok){
    // Check that the table is not missing any required columns.
    for(Int i = reqcols.nelements(); i--;){
      if(!rw.record().isDefined(reqcols[i])){
	os << LogIO::SEVERE
	   << "Required column " << reqcols[i] << " is missing."
	   << LogIO::POST;
	ok = False;// break;
      }
    }
  }
  if(!ok){
    os << name + " has an incompatible format."
       << "\nYou may want to notify the CASA system manager about it."
       << LogIO::EXCEPTION;
    return False;
  }

  // Now look for optional columns.
  Vector<String> foundoptcols;
  uInt noptcolsfound = 0;
  for(uInt i = 0; i < optcols.nelements(); ++i){
    if(rw.record().isDefined(optcols[i])){
      ++noptcolsfound;
      foundoptcols.resize(noptcolsfound, true);
      foundoptcols[noptcolsfound - 1] = optcols[i];
    }
  }

  // Together these are equiv. to optcols.assign(foundoptcols), but I find this clearer.
  optcols.resize(noptcolsfound);
  optcols = foundoptcols;
  
  table = tab;
  kws = ks;
  row = rw;
  rfp.resize(reqcols.nelements() + noptcolsfound);
  for(uInt i = 0; i < reqcols.nelements(); ++i)
    rfp[i] = RORecordFieldPtr<Double>(row.record(), reqcols[i]);
  for(uInt i = 0; i < noptcolsfound; ++i)
    rfp[reqcols.nelements() + i] = RORecordFieldPtr<Double>(row.record(), optcols[i]);
  return True;
}

// Helper function for getTable().
Bool MeasIERS::findTab(Table& tab, const Table *tabin, const String &rc,
		       const String &dir, const String &name)
{
  Bool ok = true;
  LogIO os(LogOrigin("MeasIERS", "findTab", WHERE));
  
  if(!tabin){				// No table object given: search name
    String ldir;
    Vector<String> searched;

    if(name[0] == '/'){			// Absolute path given.
      ldir = "";
    }
    else{
      const String path[2] = {
	"/ephemerides/",
	"/geodetic/"
      };

      if (Aipsrc::find(ldir, rc)){
	ldir += '/';
	searched.resize(searched.nelements() + 1, True);
	searched[searched.nelements() - 1] = ldir;
      }
      else{
	String udir;

	if(!dir.empty())
	  udir = dir + '/';
	ldir = "./";
	searched.resize(searched.nelements()+1, True);
	searched[searched.nelements()-1] = ldir;
	if (!Table::isReadable(ldir + name)) {
	  ldir = "./data/";
	  searched.resize(searched.nelements() + 1, True);
	  searched[searched.nelements()-1] = ldir;
	  if (!Table::isReadable(ldir + name)) {
	    Bool found = False;
	    String mdir;
	    if (Aipsrc::find(mdir, "measures.directory")) {
	      mdir.trim();
	      Path mpath = Path(mdir);
	      mpath.append(udir);
	      ldir = mpath.absoluteName()+"/";
	      searched.resize(searched.nelements()+1, True);
	      searched[searched.nelements()-1] = ldir;                        
	      if  (Table::isReadable(ldir+name)) {
		found = True;
	      }
	    }
	    if (!found) {
	      for (Int i=0; i<2; i++) {
		ldir = Aipsrc::aipsHome() + "/data/" + path[i];
		searched.resize(searched.nelements()+1, True);
		searched[searched.nelements()-1] = ldir;
		if (Table::isReadable(ldir + name)) {
		  found = True;
		  break;
		}
		ldir = Aipsrc::aipsRoot() + "/data/" + path[i];
		searched.resize(searched.nelements()+1, True);
		searched[searched.nelements()-1] = ldir;
		if (Table::isReadable(ldir + name)) {
		  found = True;
		  break;
		}
		Path cdatapath(String(CASADATA));
		ldir = cdatapath.absoluteName() + path[i];
		searched.resize(searched.nelements() + 1, True);
		searched[searched.nelements() - 1] = ldir;
		if (Table::isReadable(ldir + name)) {
		  found = True;
		  break;
		}
		ldir = cdatapath.absoluteName() + "/share/casacore/data/" \
		  + path[i];
		ldir = cdatapath.absoluteName() + path[i];
		searched.resize(searched.nelements() + 1, True);
		searched[searched.nelements() - 1] = ldir;
		if (Table::isReadable(ldir + name)) {
		  found = True;
		  break;
		}              
	      }
	    }
	  }
	}
      }
    }
    if(!Table::isReadable(ldir + name)){
      os << LogIO::WARN <<
	String("Requested data table ") << name <<
	String(" cannot be found in the searched directories:\n");
      for(uInt i = 0; i < searched.nelements(); ++i)
	os << searched[i] << "\n";
      os << LogIO::POST;
      return False;
    }
    tab = Table(ldir + name);
  }
  else
    tab = *tabin;

  return ok;
}

// Helper function for getTable().
Bool MeasIERS::handle_keywords(Double &dt, String &vs, const TableRecord& ks,
			       const Table& tab)
{
  LogIO os(LogOrigin("MeasIERS", "handle_keywords", WHERE));
  Bool ok = true;
  
  if(!ks.isDefined("VS_DATE") || !ks.isDefined("VS_VERSION") ||
     !ks.isDefined("VS_CREATE") || !ks.isDefined("VS_TYPE") ||
     (tab.tableInfo().type() != String("IERS"))){
    ok = False;
    os << LogIO::DEBUG1
       << "ks.isDefined(VS_DATE) " << ks.isDefined("VS_DATE")
       << "\nks.isDefined(VS_VERSION) " << ks.isDefined("VS_VERSION")
       << "\nks.isDefined(VS_CREATE) " << ks.isDefined("VS_CREATE")
       << "\nks.isDefined(VS_TYPE) " << ks.isDefined("VS_TYPE")
       << "\ntab.tableInfo().type() " << tab.tableInfo().type()
       << LogIO::POST;
  }
  if (ok) {
    Quantity ldt;
    if (MVTime::read(ldt, ks.asString("VS_DATE"))) {
      dt = MVTime(ldt);
      vs = ks.asString("VS_VERSION");
    } else {
      ok = False;
    }
  }
  return ok;
}

Bool MeasIERS::fillMeas(MeasIERS::Files which, Double utf) {
  Int ut = ifloor(utf);
  if (ut < mjd0[which] + 1 || ut >= mjdl[which]) {
    return False;
  }
  if (utf >= ldat[which][0] &&
      utf <= ldat[which+N_Files][0]) {
    // Already there
  } else {
    if (utf >= ldat[which+N_Files][0] && 
	utf <= ldat[which+N_Files][0] + 1) {
      // Shift one
      for (Int i=0; i<MeasIERS::N_Types; i++) {
	ldat[which][i] = ldat[which+N_Files][i];
      }
      // For end points
      ut = Int(ldat[which][0]);
    } else {
      // Read first line
      row[which].get(ut-mjd0[which]-1);
      for (Int i=0; i<MeasIERS::N_Types; i++) {
	ldat[which][i] = *(rfp[which][i]);
      }
    }
    // Read second line
    row[which].get(Int(ldat[which][0])-mjd0[which]);
    for (Int i=0; i<MeasIERS::N_Types; i++) {
      ldat[which+N_Files][i] = *(rfp[which][i]);
    }
    if (ldat[which][0] != ut || ldat[which+N_Files][0] != ut+1) {
      LogIO os(LogOrigin("MeasIERS",
			 String("fillMeas(MeasIERS::Files, Double)"),
			 WHERE));
      os <<
	String("The IERS table ") + tp[which] +
	" has been corrupted: regenerate" <<
	LogIO::EXCEPTION;
    }
  }
  return True;
}
  
volatile Bool MeasIERS::measFlag[MeasIERS::N_Files] = {True, True};
Double MeasIERS::dateNow = 0.0;
Table MeasIERS::t[MeasIERS::N_Files];
ROTableRow MeasIERS::row[MeasIERS::N_Files];
RORecordFieldPtr<Double> MeasIERS::rfp[MeasIERS::N_Files][MeasIERS::N_Types];
Int MeasIERS::mjd0[MeasIERS::N_Files] = {0, 0};
Int MeasIERS::mjdl[MeasIERS::N_Files] = {0, 0};
Double MeasIERS::ldat[2*MeasIERS::N_Files][MeasIERS::N_Types];
Bool MeasIERS::msgDone = False;
const String MeasIERS::tp[MeasIERS::N_Files] = {"IERSeop97", "IERSpredict"};
uInt MeasIERS::sizeNote = 0;
uInt MeasIERS::nNote = 0;
MeasIERS::CLOSEFUN *MeasIERS::toclose = 0;
Mutex MeasIERS::theirMutex;

} //# NAMESPACE CASA - END
