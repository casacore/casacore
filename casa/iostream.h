//# iostream.h: Interim solution for standard/nonstandard system iostream
//# Copyright (C) 2001,2002
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef CASA_IOSTREAM_H
#define CASA_IOSTREAM_H

//# Define the C standard C++ include file. 
//# This is an interim solution to cater for the SGI non-existence of
//# them (e.g. <cstring>)

//# Make sure any special macros are set
#include <casacore/casa/aips.h>
#include <iostream>

namespace casacore { //# NAMESPACE CASACORE - BEGIN

using std::istream;
using std::ostream;
using std::iostream;
using std::cin;
using std::cout;
using std::cerr;
using std::endl;
using std::flush;
using std::ws;

using std::ios;
using std::streampos;
using std::streamoff;
using std::streamsize;
using std::dec;
using std::hex;
using std::oct;

using std::internal;
using std::left;
using std::right;
using std::fixed;
using std::scientific;

using std::boolalpha;
using std::noboolalpha;
using std::showbase;
using std::noshowbase;
using std::showpoint;
using std::noshowpoint;
using std::showpos;
using std::noshowpos;
using std::skipws;
using std::noskipws;
using std::uppercase;
using std::nouppercase;

} //# NAMESPACE CASACORE - END

#endif
