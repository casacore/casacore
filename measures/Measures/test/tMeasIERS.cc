//# tMeasIERS.cc: This program tests the MeasIERS functions
//# Copyright (C) 2014
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This program is free software; you can redistribute it and/or modify it
//# under the terms of the GNU General Public License as published by the Free
//# Software Foundation; either version 2 of the License, or (at your option)
//# any later version.
//#
//# This program is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
//# more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with this program; if not, write to the Free Software Foundation, Inc.,
//# 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id: tMeasJPL.cc 21401 2013-11-28 11:39:17Z gervandiepen $

//# Includes
#include <casacore/casa/aips.h>
#include <casacore/measures/Measures/MeasIERS.h>
#include <casacore/casa/Quanta/MVTime.h>
#include <casacore/casa/Utilities/Assert.h>
#include <casacore/casa/iostream.h>
#include <casacore/casa/iomanip.h>

#include <casacore/casa/namespace.h>

void getX (double date)
{
  double result;
  bool sts = MeasIERS::get(result, MeasIERS::MEASURED, MeasIERS::X, date);
  cout <<"MEASURED  "<<date<< ' '<< "X: " << result << ' '<<sts<<endl;
  sts = MeasIERS::get(result, MeasIERS::PREDICTED, MeasIERS::X, date);
  cout <<"PREDICTED "<<date<< ' '<< "X: " << result << ' '<<sts<<endl;
}

int main()
{
  try {

    cout << "Test measure class MeasIERS" << endl;
    cout << "---------------------------" << endl;

    // Get a value for various epochs.

    cout << setprecision(9);
    Vector<Double> val(6);

    double date= 51116;
    for (int i=0; i<3; ++i) {
      getX (date);
      date += 0.5;
    }
    getX (37660);
    getX (37665);
    getX (55000);
    getX (55809);
    getX (600000);


    // Test for handling of leap seconds (CAS-7984)

    Double startMJD = 57202; // 2015-06-29T00:00:00
    Double oneHour = 1.0/24.;
    for (Int i = 0; i < 72; i++) {
      const Double mjd = startMJD + i * oneHour;
      const MVTime now(mjd);
      Double dUT;
      Bool rval = MeasIERS::get(dUT, MeasIERS::PREDICTED, MeasIERS::dUT1, mjd);
      if(!rval){
	cout << "MeasIERS::get returned False for PREDICTED, dUT1, mjd " << mjd << endl;
	return 2;
      }
      cout << now.string(casa::MVTime::YMD) 
	   << " " << setprecision(3) << fixed << mjd
	   << " " << dUT << endl;
    } 
    

  } catch (const std::exception& x) {
    cout << x.what() << endl;
    return 1;
  } 
  
  return 0;
}
