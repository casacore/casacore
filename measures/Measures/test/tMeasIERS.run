#!/bin/sh

# Define the casarc file and create it.
cwd=`pwd`
CASARCFILES=$cwd/tMeasIERS_tmp-casarc
export CASARCFILES
cat > $CASARCFILES << eof
measures.ierseop97.directory: $cwd/tMeasIERS_tmp-data/geodetic
measures.ierspredict.directory: $cwd/tMeasIERS_tmp-data/geodetic
eof

# Untar the IERS tables
rm -rf tMeasIERS_tmp-data
tar zxf tMeasIERS.in_tgz

# Run the test program
$casa_checktool ./tMeasIERS
