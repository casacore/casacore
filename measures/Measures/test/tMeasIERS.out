Test measure class MeasIERS
---------------------------
MEASURED  51116 X: 0.15712 1
PREDICTED 51116 X: 0 0
MEASURED  51116.5 X: 0.157685 1
PREDICTED 51116.5 X: 0 0
MEASURED  51117 X: 0.15825 1
PREDICTED 51117 X: 0 0
MEASURED  37660 X: 0 0
PREDICTED 37660 X: 0 0
MEASURED  37665 X: -0.0127 1
PREDICTED 37665 X: 0 0
MEASURED  55000 X: 0.091192 1
PREDICTED 55000 X: 0.09119 1
MEASURED  55809 X: 0.166538 1
PREDICTED 55809 X: 0.16653 1
MEASURED  600000 X: 0 0
PREDICTED 600000 X: 0 0
2015/06/29/00:00:00 57202.000 -0.675
2015/06/29/01:00:00 57202.042 -0.675
2015/06/29/02:00:00 57202.083 -0.675
2015/06/29/03:00:00 57202.125 -0.675
2015/06/29/04:00:00 57202.167 -0.676
2015/06/29/05:00:00 57202.208 -0.676
2015/06/29/06:00:00 57202.250 -0.676
2015/06/29/07:00:00 57202.292 -0.676
2015/06/29/08:00:00 57202.333 -0.676
2015/06/29/09:00:00 57202.375 -0.676
2015/06/29/10:00:00 57202.417 -0.676
2015/06/29/11:00:00 57202.458 -0.676
2015/06/29/12:00:00 57202.500 -0.676
2015/06/29/13:00:00 57202.542 -0.676
2015/06/29/14:00:00 57202.583 -0.676
2015/06/29/15:00:00 57202.625 -0.676
2015/06/29/16:00:00 57202.667 -0.676
2015/06/29/17:00:00 57202.708 -0.676
2015/06/29/18:00:00 57202.750 -0.676
2015/06/29/19:00:00 57202.792 -0.676
2015/06/29/20:00:00 57202.833 -0.676
2015/06/29/21:00:00 57202.875 -0.676
2015/06/29/22:00:00 57202.917 -0.676
2015/06/29/23:00:00 57202.958 -0.676
2015/06/30/00:00:00 57203.000 -0.676
2015/06/30/01:00:00 57203.042 -0.676
2015/06/30/02:00:00 57203.083 -0.676
2015/06/30/03:00:00 57203.125 -0.676
2015/06/30/04:00:00 57203.167 -0.676
2015/06/30/05:00:00 57203.208 -0.676
2015/06/30/06:00:00 57203.250 -0.676
2015/06/30/07:00:00 57203.292 -0.676
2015/06/30/08:00:00 57203.333 -0.676
2015/06/30/09:00:00 57203.375 -0.676
2015/06/30/10:00:00 57203.417 -0.676
2015/06/30/11:00:00 57203.458 -0.676
2015/06/30/12:00:00 57203.500 -0.676
2015/06/30/13:00:00 57203.542 -0.676
2015/06/30/14:00:00 57203.583 -0.676
2015/06/30/15:00:00 57203.625 -0.676
2015/06/30/16:00:00 57203.667 -0.677
2015/06/30/17:00:00 57203.708 -0.677
2015/06/30/18:00:00 57203.750 -0.677
2015/06/30/19:00:00 57203.792 -0.677
2015/06/30/20:00:00 57203.833 -0.677
2015/06/30/21:00:00 57203.875 -0.677
2015/06/30/22:00:00 57203.917 -0.677
2015/06/30/23:00:00 57203.958 -0.677
2015/07/01/00:00:00 57204.000 0.323
2015/07/01/01:00:00 57204.042 0.323
2015/07/01/02:00:00 57204.083 0.323
2015/07/01/03:00:00 57204.125 0.323
2015/07/01/04:00:00 57204.167 0.323
2015/07/01/05:00:00 57204.208 0.323
2015/07/01/06:00:00 57204.250 0.323
2015/07/01/07:00:00 57204.292 0.323
2015/07/01/08:00:00 57204.333 0.323
2015/07/01/09:00:00 57204.375 0.323
2015/07/01/10:00:00 57204.417 0.323
2015/07/01/11:00:00 57204.458 0.323
2015/07/01/12:00:00 57204.500 0.323
2015/07/01/13:00:00 57204.542 0.323
2015/07/01/14:00:00 57204.583 0.323
2015/07/01/15:00:00 57204.625 0.323
2015/07/01/16:00:00 57204.667 0.323
2015/07/01/17:00:00 57204.708 0.323
2015/07/01/18:00:00 57204.750 0.323
2015/07/01/19:00:00 57204.792 0.323
2015/07/01/20:00:00 57204.833 0.323
2015/07/01/21:00:00 57204.875 0.323
2015/07/01/22:00:00 57204.917 0.323
2015/07/01/23:00:00 57204.958 0.323
